sql/sql_class.h

```diff
@@ -1588,20 +1590,29 @@ class Bloom_filter {
 
   typedef std::bitset<SIZE> Bit_set;
 
-  /** The bit set, which is allocated in a MEM_ROOT */
-  Bit_set *bit_set;
+  typedef boost::shared_ptr<Bit_set> Bit_set_ptr;
+
+  Bit_set_ptr bit_set;
+  // Protects the bit set from concurrent assignment
+  mysql_mutex_t LOCK_bit_set;
 
   // Non-copyable
   Bloom_filter (const Bloom_filter &);
   Bloom_filter & operator = (const Bloom_filter &);
 
  public:
-  Bloom_filter() : bit_set(NULL) {}
+  Bloom_filter() {
+    mysql_mutex_init(key_LOCK_bloom_filter,
+                   &LOCK_bit_set, MY_MUTEX_INIT_FAST);
+  }
+
+  ~Bloom_filter() {
+    mysql_mutex_destroy(&LOCK_bit_set);
+  }
 
   void clear()
   {
-    if (bit_set != NULL)
-      bit_set->reset();
+    bit_set.reset();
```

Comment:
Here, I would also protect this with a mutex.
```suggestion
    mysql_mutex_lock(&LOCK_bit_set);
    Bit_set_ptr local_bit_set = my_atomic_loadptr(&bit_set);
    delete local_bit_set;
    my_atomic_storeptr(&bit_set, NULL);
    mysql_mutex_unlock(&LOCK_bit_set);
```